packages/audiofileplayer/lib/audiofileplayer.dart

```diff
@@ -542,20 +721,195 @@ class Audio with WidgetsBindingObserver {
   void didChangeAppLifecycleState(AppLifecycleState state) {
     if (state == AppLifecycleState.paused) {
       _appPaused = true;
-      if (_playing && !shouldPlayWhileAppPaused) {
+      if (_playing && !_playInBackground) {
         _pauseNative();
       }
     } else if (state == AppLifecycleState.resumed) {
       _appPaused = false;
-      if (_playing && !shouldPlayWhileAppPaused) {
+      if (_playing && !_playInBackground) {
         _playNative(false, _endpointSeconds);
       }
     }
   }
 
+  /// Inform the OS's background audio system about the playback state; used to
+  /// set the progress bar in lockscreen/notification.
+  ///
+  /// When using background audio, this should be called every time the playback
+  /// starts/stops, or there is discontinuity in playback progress (i.e. a seek
+  /// or loop restart). It should not be called regularly (i.e. do not call it
+  /// every second).
+  static void setPlaybackState(bool isPlaying, double positionSeconds) async {
```

Comment:
This method is communicating what the OS (e.g. the lock screen/control center/notification) is displaying; it is similar to the other tell-the-OS-what-to-display methods (like setMetadata, etc). It is not tied to a particular instance of Audio, so doesn't make sense to be an instance method.

Reply:
I looked over the entire file to get a better picture of how this class is structured, specifically the static methods.

The static constructors are fine. That's due to Dart's inability to support async factory methods. A MockAudio class would solve testability problem here.

Like you said, any of the other static methods (setMetadata) are not tied to an instance of `Audio`. IMO, they belong to a different semantic entity altogether. So, I suggest to pull these into a different class `BackgroundAudioService` which is a singleton service and contains methods to configure the service. This will allow apps to create mocks of this service and test the widgets that call `setMetadata`, `setPlaybackState` etc.

What do you think?

Reply:
I'm open to it, but:

- I'm ok with a separate class/file to hold the static methods. But the concept of a 'service' (in terms of how an app communicates with the OS's background audio system) doesn't translate into Dart/Flutter land (it's used in Android, but not on iOS), and so I'd like to hide that implementation detail from naming this new class/file.

- I'd prefer to avoid adding code complexity for testing purposes, by making a Singleton instance where static methods work more elegantly. For testing purposes, can't callers do exactly what Audio instances do (i.e. in test/audiofileplayer_test.dart)? i.e. use setMockMethodCallHandler() ? Call Audio.setMetadata() normally, and listen to what gets sent via the channel.
If it's purely about dart-mockito not being able to stub static methods, I don't want to work around that limitation in my production code; the user can decide how they want to work around that in their tests (e.g. they could wrap the static call within an instance method of a wrapper and mock/stub that wrapper, i.e. create their own singleton)

Let me know what you think...

Reply:
The new [federated plugins design](http://flutter.dev/go/federated-plugins) encourages platform-specific APIs if appropriate. The main reason is that we should not assume all Flutter developers target all platforms. Flutter is popular on iOS and Android at the moment but the vision is to expand to many more. 

So, if these APIs are not used on iOS at all, I actually would want to move them to a properly namespaced file even more. I have seen (an even worse) practice of `assert(Platform.isAndroid);` in other APIs which I think is even more confusing for developers. At least, with the properly named file, users know not to call these APIs if they are not targeting Android.

You can name this file anything you want btw, I am not hung up on "service".

As far as the static methods go.. that's a matter of stylistic opinion. However, there are real costs to your users:
- Users of your plugin should never need to do setMockMethodCallHandler on your plugin. Your method call handler is internal implementation that only your tests are supposed to understand.
- Indeed, users will do exactly what you are describing (we have examples of this in google3). That ends up being much less elegant since they are now creating their own wrapper + a singleton on top of a static API. 
- If the users want to code against a "static" API (and write no tests), they can still do that with a single extra line which is less costly than what's described above:

```
const backgroundAudioService = BackgroundAudioService();
```
- It gives you slightly better flexibility. In the future, if you wanted to store some state for these APIs, you might be able to get away with it w/o introducing a breaking change. We have seen a couple of examples of this (and is also the same reason why I don't recommend getters with a heavy implementation).

Ultimately, this ends up being about which best practices we are encouraging. In google3, we lean towards enabling better/cleaner testing practices. Sometimes, that means writing code in a specific way.

Reply:
Ok, thanks for the detail, I've unresolved this.

I'll update the static methods into Singletons with instance methods.

In the meantime, I've looked at the federated plugin doc, but I'm still a little unclear what you are suggesting for this PR. 

* There is only one method which is Android-only: `setAndroidMediaButtons()` (all the other static background audio methods, e.g. setMetadata(), setPlaybackState(), setSupportedActions() are necessary on both iOS and Android)

* There is only one method which is iOS-only: `setIosAudioCategory()`.

So are you suggesting
* Dart:
  * class Audio
  * class AndroidAudio (singleton, one method)
  * class IosAudio (singleton, one method)
* Android:
  * class AudioPlugin
  * class AndroidAudioPlugin
* iOS
  * class AudioPlugin
  * class IosAudioPlugin

?

Reply:
Each platform implementation can be a single class (single class handles both common and platform-specific functionality). I gave you a link to the federated plugins doc because you said the static methods are Android specific functionality. Federated plugins discussion is actually orthogonal to static vs instance method discussion.

Eventually, this is what's going to happen:

- Dart
  - class Audio (common functionality)
  - class AudioConfiguration/Service/WhatHaveYou (singleton, "static" common methods)
  - class AndroidAudio (Android-specific functionality, not sure if this needs to exist?)
  - class AndroidAudioConfiguration (Android-specific functionality, static methods?)
  - class IosAudio (iOS-specific functionality, not sure if this needs to exist?)
  - class IosAudioConfiguration (iOS-specific functionality, static methods?)
- Android
  - class AudioPlugin
- iOS
  - class AudioPlugin

We are not requiring all plugins to migrate to this yet so you don't have to do it in this PR. Let's keep the Android/iOS/common split out of scope.

For now, let's just create a new AudioConfiguration class which hosts the currently static methods and make them instance methods. That should be sufficient. After that, this PR LGTM.

Reply:
SG, moved static methods into new singleton class AudioSystem, PTAL, thanks.